core/trino-main/src/main/java/io/trino/operator/exchange/ScaleWriterPartitioningExchanger.java

```diff
@@ -0,0 +1,154 @@
+        implements LocalExchanger
+{
+    private final List<Consumer<PageReference>> buffers;
+    private final LocalExchangeMemoryManager memoryManager;
+    private final long maxBufferedBytes;
+    private final Function<Page, Page> partitionedPagePreparer;
+    private final PartitionFunction partitionFunction;
+    private final UniformPartitionRebalancer partitionRebalancer;
+    private final PageReleasedListener onPageReleased;
+
+    private final IntArrayList[] partitionAssignments;
+    private final IntArrayList[] writerAssignments;
+
+    public ScaleWriterPartitioningExchanger(
+            List<Consumer<PageReference>> buffers,
+            LocalExchangeMemoryManager memoryManager,
+            long maxBufferedBytes,
+            Function<Page, Page> partitionedPagePreparer,
+            PartitionFunction partitionFunction,
+            int partitionCount,
+            UniformPartitionRebalancer partitionRebalancer)
+    {
+        this.buffers = requireNonNull(buffers, "buffers is null");
+        this.memoryManager = requireNonNull(memoryManager, "memoryManager is null");
+        this.maxBufferedBytes = maxBufferedBytes;
+        this.partitionedPagePreparer = requireNonNull(partitionedPagePreparer, "partitionedPagePreparer is null");
+        this.partitionFunction = requireNonNull(partitionFunction, "partitionFunction is null");
+        this.partitionRebalancer = requireNonNull(partitionRebalancer, "partitionRebalancer is null");
+        this.onPageReleased = forLocalExchangeMemoryManager(memoryManager);
+
+        // Initialize writerAssignments with the buffer size
+        writerAssignments = new IntArrayList[buffers.size()];
+        for (int i = 0; i < writerAssignments.length; i++) {
+            writerAssignments[i] = new IntArrayList();
+        }
+
+        // Initialize partitionAssignments and partitionScalingInfos with the artificial partition limit.
+        partitionAssignments = new IntArrayList[partitionCount];
+        for (int i = 0; i < partitionAssignments.length; i++) {
+            partitionAssignments[i] = new IntArrayList();
+        }
+    }
+
+    @Override
+    public void accept(Page page)
+    {
+        Page partitionPage = partitionedPagePreparer.apply(page);
+
+        // Assign each row to a partition which limits to the partitionCount. If there are more physical partitions than
+        // this artificial partition limit, then it is possible that multiple partitions will get assigned the same
+        // bucket id. Thus, multiple partitions will be scaled together since we track partition physicalWrittenBytes
+        // using the artificial limit (partitionCount). The assignments lists are all expected to cleared by the
+        // previous iterations.
+        for (int position = 0; position < partitionPage.getPositionCount(); position++) {
+            int partition = partitionFunction.getPartition(partitionPage, position);
+            partitionAssignments[partition].add(position);
```

Comment:
No, this isn't required since we are creating a separate instance of `ScaleWriterPartitioningExchanger` for each operator. The only thing that needs to be thread-safe is `UniformPartitionRebalancer` since it keeps the partition scaling state.